// hesperus/dist/structures/command/SlashCommandChoice.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlashCommandChoice = void 0;
/**
 * This class represents slash command choice.
 */
class SlashCommandChoice {
    /**
     * @param {SlashCommandChoiceOptions} _options Choice's options.
     */
    constructor(_options) {
        this._options = _options;
    }
    /**
     * Name of this slash command choice.
     *
     * @return {string}
     */
    get name() {
        return this._options.name;
    }
    /**
     * Localizations for name of this slash command choice.
     *
     * @return {LocalizationMap | undefined}
     */
    get nameLocalizations() {
        return this._options.nameLocalizations;
    }
    /**
     * Value of this slash command choice.
     *
     * @return {string | number}
     */
    get value() {
        return this._options.value;
    }
    /**
     * This method allows you to convert this class to suitable
     * (for Discord) JSON.
     *
     * @return {object} Converted class.
     */
    toJSON() {
        return {
            name: this.name,
            value: this.value,
            ...(this.nameLocalizations && { name_localizations: this.nameLocalizations }),
        };
    }
}
exports.SlashCommandChoice = SlashCommandChoice;
