// hesperus/dist/structures/command/SlashCommandOption.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlashCommandOption = exports.ChannelTypes = exports.SlashCommandOptionType = void 0;
/**
 * This enum represents slash command's option type.
 */
var SlashCommandOptionType;
(function (SlashCommandOptionType) {
    SlashCommandOptionType[SlashCommandOptionType["SUB_COMMAND"] = 1] = "SUB_COMMAND";
    SlashCommandOptionType[SlashCommandOptionType["SUB_COMMAND_GROUP"] = 2] = "SUB_COMMAND_GROUP";
    SlashCommandOptionType[SlashCommandOptionType["STRING"] = 3] = "STRING";
    SlashCommandOptionType[SlashCommandOptionType["INTEGER"] = 4] = "INTEGER";
    SlashCommandOptionType[SlashCommandOptionType["BOOLEAN"] = 5] = "BOOLEAN";
    SlashCommandOptionType[SlashCommandOptionType["USER"] = 6] = "USER";
    SlashCommandOptionType[SlashCommandOptionType["CHANNEL"] = 7] = "CHANNEL";
    SlashCommandOptionType[SlashCommandOptionType["ROLE"] = 8] = "ROLE";
    SlashCommandOptionType[SlashCommandOptionType["MENTIONABLE"] = 9] = "MENTIONABLE";
    SlashCommandOptionType[SlashCommandOptionType["NUMBER"] = 10] = "NUMBER";
    SlashCommandOptionType[SlashCommandOptionType["ATTACHMENT"] = 11] = "ATTACHMENT";
})(SlashCommandOptionType = exports.SlashCommandOptionType || (exports.SlashCommandOptionType = {}));
/**
 * This enum represents slash command's channel types.
 */
var ChannelTypes;
(function (ChannelTypes) {
    ChannelTypes[ChannelTypes["GUILD_TEXT"] = 0] = "GUILD_TEXT";
    ChannelTypes[ChannelTypes["DM"] = 1] = "DM";
    ChannelTypes[ChannelTypes["GUILD_VOICE"] = 2] = "GUILD_VOICE";
    ChannelTypes[ChannelTypes["GROUP_DM"] = 3] = "GROUP_DM";
    ChannelTypes[ChannelTypes["GUILD_CATEGORY"] = 4] = "GUILD_CATEGORY";
    ChannelTypes[ChannelTypes["GUILD_NEWS"] = 5] = "GUILD_NEWS";
    ChannelTypes[ChannelTypes["GUILD_NEWS_THREAD"] = 10] = "GUILD_NEWS_THREAD";
    ChannelTypes[ChannelTypes["GUILD_PUBLIC_THREAD"] = 11] = "GUILD_PUBLIC_THREAD";
    ChannelTypes[ChannelTypes["GUILD_PRIVATE_THREAD"] = 12] = "GUILD_PRIVATE_THREAD";
    ChannelTypes[ChannelTypes["GUILD_STAGE_VOICE"] = 13] = "GUILD_STAGE_VOICE";
    ChannelTypes[ChannelTypes["GUILD_DIRECTORY"] = 14] = "GUILD_DIRECTORY";
    ChannelTypes[ChannelTypes["GUILD_FORUM"] = 15] = "GUILD_FORUM";
})(ChannelTypes = exports.ChannelTypes || (exports.ChannelTypes = {}));
/**
 * This class represents option for command, that can be added to it.
 */
class SlashCommandOption {
    /**
     * @param {SlashCommandOptionOptions} _options Option's options.
     */
    constructor(_options) {
        this._options = _options;
        // TODO: add checks for `min_value`, `max_value` and `autocomplete`
    }
    /**
     * Type of this slash command option.
     *
     * @return {SlashCommandOptionType}
     */
    get type() {
        return this._options.type;
    }
    /**
     * Name of this slash command option.
     *
     * @return {string}
     */
    get name() {
        return this._options.name;
    }
    /**
     * Localizations for name of this slash command option.
     *
     * @return {LocalizationMap | undefined}
     */
    get nameLocalizations() {
        return this._options.nameLocalizations;
    }
    /**
     * Descriptions of this slash command option.
     *
     * @return {string}
     */
    get description() {
        return this._options.description;
    }
    /**
     * Localizations for description of this slash command option.
     *
     * @return {LocalizationMap | undefined}
     */
    get descriptionLocalizations() {
        return this._options.descriptionLocalizations;
    }
    /**
     * Whether this slash command option is required.
     *
     * @return {boolean | undefined}
     */
    get required() {
        return this._options.required;
    }
    /**
     * Choices of this slash command option.
     *
     * @return {SlashCommandChoice[] | undefined}
     */
    get choices() {
        return this._options.choices;
    }
    /**
     * Types of channel of this slash command option (only for `CHANNEL` type).
     *
     * @return {ChannelTypes[] | undefined}
     */
    get channelTypes() {
        return this._options.channelTypes;
    }
    /**
     * Minimal value of this slash command option
     * (only for `INTEGER` and `NUMBER` types).
     *
     * @return {number | undefined}
     */
    get minValue() {
        return this._options.minValue;
    }
    /**
     * Maximal value of this slash command option
     * (only for `INTEGER` and `NUMBER` types).
     *
     * @return {number | undefined}
     */
    get maxValue() {
        return this._options.maxValue;
    }
    /**
     * Whether we should enable autocomplete of this slash command option.
     *
     * @return {boolean | undefined}
     */
    get autocomplete() {
        return this._options.autocomplete;
    }
    /**
     * This method allows you to convert this class to suitable
     * (for Discord) JSON.
     *
     * @return {object} Converted class.
     */
    toJSON() {
        return {
            type: this.type,
            name: this.name,
            description: this.description,
            ...(this.required && { required: this.required }),
            ...(this.nameLocalizations && { name_localizations: this.nameLocalizations }),
            ...(this.descriptionLocalizations && { description_localizations: this.descriptionLocalizations }),
            ...(this.choices && { choices: this.choices.map((choice) => choice.toJSON()) }),
            ...(this.channelTypes && { channel_types: this.channelTypes }),
            ...(this.minValue && { min_value: this.minValue }),
            ...(this.maxValue && { max_value: this.maxValue }),
            ...(this.autocomplete && { autocomplete: this.autocomplete }),
        };
    }
}
exports.SlashCommandOption = SlashCommandOption;
