// hesperus/dist/structures/command/SlashSubCommandGroup.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.SlashSubCommandGroup = void 0;
/**
 * This class represents slash sub command's group.
 */
class SlashSubCommandGroup {
    /**
     * @param {SlashSubCommandGroupOptions} _options Options for this group.
     */
    constructor(_options) {
        this._options = _options;
    }
    /**
     * Name of this subcommand group.
     *
     * @return {string}
     */
    get name() {
        return this._options.name;
    }
    /**
     * Localizations for name of this subcommand group.
     *
     * @return {LocalizationMap | undefined}
     */
    get nameLocalizations() {
        return this._options.nameLocalizations;
    }
    /**
     * Description of this subcommand group.
     *
     * @return {string}
     */
    get description() {
        return this._options.description;
    }
    /**
     * Localizations for description of this subcommand group.
     *
     * @return {LocalizationMap | undefined}
     */
    get descriptionLocalizations() {
        return this._options.descriptionLocalizations;
    }
    /**
     * Commands of this subcommand group.
     *
     * @return {SlashCommand[]}
     */
    get commands() {
        return this._options.commands;
    }
    /**
     * This method allows you to convert this class to suitable
     * (for Discord) JSON.
     *
     * @return {object} Converted class.
     */
    toJSON() {
        return {
            name: this.name,
            description: this.description,
            type: 2,
            ...(this.commands && { options: this.commands.map((command) => command.toJSON()) }),
            ...(this.nameLocalizations && { name_localizations: this.nameLocalizations }),
            ...(this.descriptionLocalizations && { description_localizations: this.descriptionLocalizations }),
        };
    }
}
exports.SlashSubCommandGroup = SlashSubCommandGroup;
